### ISSUE TITLE
Implement @nnotations, dependency injection, fragments & extensions

### ISSUE THREAD

**Issue by as-shura** [BlackBeltTechnology/pandino#102]
Greetings I find this project amazing and would like to propose some development directions.

there are already other projects that have some features already implemented and I will list them as follow:

- https://github.com/lorry2018/minimajs
- https://github.com/cescoffier/hubu/

For the documentation concerning @nnotations:

https://docs.osgi.org/javadoc/osgi.cmpn/7.0.0/org/osgi/service/component/annotations/Component.html

If you would like to have an already usable framework that has @nnotations & dependency injection I recommend you check **Inversify**. It is already used by the Eclipse Theia project and have extensions already implemented for UI development like in Eclipse IDE.

https://inversify.io/

Maybe this can be a good roadmap.

Best Regards

**Comment by as-shura** [comment]
This could be related with this thread : https://github.com/BlackBeltTechnology/pandino/issues/12

**Comment by noherczeg** [comment]
@as-shura sorry for the late response. We are aware of inversify, but Pandino is pretty different. One of the key differences is the life cycle handling. While traditional DI tools  provide a "one-off" wiring with imports and references to classes / constants, Pandinonis xompletely decoupled, and supports the coming and going of bundles and services.

**Comment by noherczeg** [comment]
Another challenge for us is, that sadly there is not even an RFC in the ECMA standard which covers functions, and support for decorating them. We are currently focusing on supporting functional programming.

**Comment by as-shura** [comment]
> Another challenge for us is, that sadly there is not even an RFC in the ECMA standard which covers functions, and support for decorating them. We are currently focusing on supporting functional programming.

There is such thing already like in Actionscript MetaDatas you have some similar mechanisms in TypeScript with Decorators/Annotations alike Java.

https://www.beyondjava.net/typescript-and-es2016-decorators-vs-java-annotations

**Comment by noherczeg** [comment]
Sorry there could be a misunderstanding. We know, that ES has decorators, our "problem" is that we would like to be able to implement #12 in a way, where the decorators are as light as possible.

The reason for this is, that in a "Traditional" DI app, the decorators are classes/functions and are code split into some chunk in the application code. In Pandino, bundles do not contain any runtime code related to the OSGi runtime. Everything is decoupled.

Since interfaces are not native ES citizens, we currently rely on `string` -> `interface` pairing. Although this is not error-prone if the same interface has different version (since strings are inlined into bundled code), at least we do not rely on bundled classes / functions. At least now.

Long story short, we will have a go again with #12 in the near future, but it's not a trivial topic, unless we "give up" on current architectural values.

**Comment by noherczeg** [comment]
Pandino has 0 dependencies on any third party library. All I meant by the React reference is that when it comes to plugin/module/extension development, we've been focusing on React hooks, and gathering feedback on such "extensions".

Given that we seem to have potential client for #12 I'll look into it right now. You peaked my interest ;)

**Comment by as-salam** [comment]
How is it going concerning this feature?
Is this project dead or still alive?

Kind regards

**Comment by fayakun-it-consulting** [comment]
Greetings to @noherczeg, @as-salam, @as-shura and the whole Pandino team!

Let me introduce myself, I am Rama from Fayakun IT Consulting

We have discovered Pandino framework while browsing the Eclipse Theia project's repository and discussions.
@as-salam started a thread about integrating Inversify into Pandino the same way iPOJO was develop to be integrated into any kind of OSGi runtime. We thought that was a really interesting approach because the functionalities offered by iPOJO improved our development and architecture plumbing greatly.

He also mentioned the possibility to integrate features and functionalities (Java like annotations) similar to what TornadoFX & DromblerFX offer for dynamic UI architecture plumbing necessary for enterprise application.

Pandino is also built around the same kind of OSGi development mindset. Although Pandino is a great framework to build our application on we have realized during our exchanges that it definitely needs on top of itself something like iPOJO. Due to the registration and declaration for the incoming services which are at the moment handled all manually by the developer itself.

Pandino right now has no way of using Java style annotations for the not functional wiring of services. In the other hand Inversify has almost all the functionalities that are present inside of iPOJO (Java) or Spring DM (Java). Since it is a **Dependency Injection** container we had the idea to integrate it by using Extender pattern offered within Pandino.

It seems that other folks coming from Java who were using Inversify developed a utility package on top of Inversify to allow a Service Oriented Component feeling like iPOJO with only a small missing feature which are the decorators.

https://github.com/webcomputing/inversify-components
  
We were wondering if you would like to be part of the brainstorming process and help us come into the thread in Theia' project?

The link for the discussion is here https://github.com/eclipse-theia/theia/discussions/16001#discussioncomment-13783414

Kind Regards

**Comment by fayakun-it-consulting** [comment]
Quick update to the previous post, we found a newer project which is getting closer to what you guys might think of.

Link : https://github.com/eclipse-theia/theia/discussions/16001#discussioncomment-13855042

**Comment by fayakun-it-consulting** [comment]
> Sorry there could be a misunderstanding. We know, that ES has decorators, our "problem" is that we would like to be able to implement [#12](https://github.com/BlackBeltTechnology/pandino/issues/12) in a way, where the decorators are as light as possible. Also: we are currently working with mostly React codebases, and since it means everything is a function, we cannot decorate React hooks / `FC`-s.
> 
> The reason for this is, that in a "Traditional" DI app, the decorators are classes/functions and are code split into some chunk in the application code. In Pandino, bundles do not contain any runtime code related to the OSGi runtime. Everything is decoupled.
> 
> Since interfaces are not native ES citizens, we currently rely on `string` -> `interface` pairing. Although this is not error-prone if the same interface has a different version (since strings are inlined into bundled code), at least we do not rely on bundled classes / functions. At least for now.
> 
> Long story short, we will have a go again with [#12](https://github.com/BlackBeltTechnology/pandino/issues/12) in the near future, but it's not a trivial topic, unless we "give up" on current architectural values.

@noherczeg I understand your concern for the feasibility of [#12](https://github.com/BlackBeltTechnology/pandino/issues/12)  knowing it is not a trivial task. But I have a good news because the folks at Apache Felix project came up with the **Apache iPOJO** project completely alleviating the pain of manual service declarations. 

If you go on their web site subprojects page at : https://felix.apache.org/documentation/subprojects.html

You have a list of really cool projects like [Service Component Runtime](https://github.com/apache/felix-dev/tree/master/scr)

<img width="994" height="597" alt="Image" src="https://github.com/user-attachments/assets/e0dd8ef2-92e2-4c3d-ba90-a68b7f1c3956" />



If you go lower you will see their retired projects in which iPOJO moved to

<img width="845" height="344" alt="Image" src="https://github.com/user-attachments/assets/0bc08b06-af3b-4d09-84d6-db95e3214c74" />



They offered 3 ways of defining service components. One of them is offering the developer plumbing & wiring annotation based DI.

https://cwiki.apache.org/confluence/display/FELIX/How+to+use+iPOJO+Annotations

I don't know if you ever heard of the **Extender Pattern** bundle development inside of OSGi?

https://toedter.com/2010/07/05/osgi-4-2-extender-pattern-and-bundletracker/

The entry point for the **iPOJO** bundle lies exactly here : https://github.com/apache/felix-dev/blob/master/ipojo/runtime/core/src/main/java/org/apache/felix/ipojo/extender/internal/Extender.java

More information about the **Extender Pattern** which a really powerful option available to **iPOJO** because it open the DI framework to receive more customization with handlers  : https://cwiki.apache.org/confluence/display/FELIX/Extender+Pattern+Handler

How complicated will it be to port this project to Pandino with AI copilot translating the code 1:1 to Typescript?

There is also another alternative already existing with very powerful features like iPOJO **DI FRAMEWORK** which is based on **Inversify** with all the disadvantages taken out from traditional **DI** frameworks following the latest cutting edge methodologies. Please have a look at : https://github.com/eclipse-theia/theia/discussions/16001#discussioncomment-13855042

They also have 2 React project templates : https://github.com/inversiland/inversiland?tab=readme-ov-file#templates

Do you think that this will resolve your worries concerning [#12](https://github.com/BlackBeltTechnology/pandino/issues/12)?

**Comment by stbischof** [comment]
I also would recommend to focus on newer declarative services spec/ the scr Implementation.

**Comment by fayakun-it-consulting** [comment]
> I also would recommend to focus on newer declarative services spec/ the scr Implementation.

Which of iPOJO & Apache Felix SCR has more features?

**Comment by stbischof** [comment]
Felix SCR is the Reference Implementation of OSGi DS.

https://docs.osgi.org/specification/osgi.cmpn/7.0.0/service.component.html

This is the mentaned/supported bundle.

**Comment by noherczeg** [comment]
@stbischof , @as-salam SCR with decorator support will be released this week in Pandino. I have a not yet pushed version ready with most use-cases, and updated docs.

This will be a 0.9.x release as per current readme roadmap, and if I can get your support for evaluation and feedback, we should be able to hit the 1.0.0 in a couple of weeks.

**Comment by noherczeg** [comment]
@as-salam inversiland is essentially an Angular-ish thing, therefore much different compared to the OSGi way. In Pandino we focus on complete de-coupling while e.g. they are referencing and relying on classes, etc.

I see no point in introducing third party libraries to achieve the core features of the OSGi spec, therefore it's completely self contained.

**Comment by as-salam** [comment]
I invited @pkriens who is one of the founding fathers of OSGi to also join the conversation. I don't know if he is aware of the existence of Pandino runtime. Maybe he might be pleased to see these threads running.

**Comment by stbischof** [comment]
@as-salam i am pretty sure that you do not need to invite more people who worked on osgi.

Lead persons of the steering committee and the specification project know that pandino exists and they are happy to see that now scr is implemented here.

**Comment by as-salam** [comment]
> @as-salam i am pretty sure that you do not need to invite more people who worked on osgi.
> 
> Lead persons of the steering committee and the specification project know that pandino exists and they are happy to see that now scr is implemented here.

:smile: Thank you, I didn't know about that! If Pandino is that well known why isn't it used a lot?

**Comment by noherczeg** [comment]
> 😄 Thank you, I didn't know about that! If Pandino is that well known why isn't it used a lot?

From our perspective we've been using it since inception, internally in our other frameworks / projects. The challenge was the fact that the official OSGi universe is so wast, we wanted to make sure whatever we end up with is stable and usable. Therefore wanted to market it more when we reach 1.0.0.

To give you some context: we are 99% React when it comes to frontend, and sadly we cannot really utilize SCR. That is the main reason why I lagged so hard with the implementation.

**Comment by noherczeg** [comment]
@as-salam , @stbischof as promised the framework has been re-implemented with SCR support:

https://github.com/BlackBeltTechnology/pandino/releases/tag/v0.9.22

**Comment by noherczeg** [comment]
Please play around with it, and provide feedback, so that we can improve even further :) If it's okay, then you can close this issue, I'll leave it to you guys.

**Comment by noherczeg** [comment]
@as-shura do you still need fragment support?

**Comment by as-salam** [comment]
> @as-shura do you still need fragment support?

Fragment support would be the cherry on the cake! With fragments we can zip an entire web App and deliver dynamic UI parts to the main UI with the CSS, HTML & images. This is how we used to do in Java OSGi.

Do you also know the extensions mechanism in the Eclipse IDE environment? I posted something similar related with iPOJO which offered such mechanism.

Let me find out

**Comment by as-salam** [comment]
> Please play around with it, and provide feedback, so that we can improve even further :) If it's okay, then you can close this issue, I'll leave it to you guys.

If you closr this issue will you receive any updates if we add other threads?

**Comment by as-salam** [comment]
You can close this thread we will follow the rest in other issues.

Good work!
